### DISCUSSION TITLE
setup_file and setup does find few issues, may be improve my understanding.

### DISCUSSION THREAD

**Discussion by aditya-malviya** [bats-core/bats-core#1022] [Q&A]
I'm working on bats and have couple of issues.

assume one.batsas my bats file. 

```
#!/usr/bin/env bats
setup_file(){
  echo "#### Test for file abc.sh is running now ####" >&3
}

setup() {
  load $HOME/test/'test_helper/batsassert/load.bash'
 load $HOME/test/'test_helper/batscore/load.bash'
  load './one.sh'
}

@test "Test successful compliance check" {
  run process_one 0.  #function call to process_one with params 0 
  [ "$status" -eq 0 ]
}
```
------------
If i move 

```
load $HOME/test/'test_helper/batsassert/load.bash'
 load $HOME/test/'test_helper/batscore/load.bash'
  load './one.sh'
  
  to 

setup_file(){
  echo "#### Test for file abc.sh is running now ####" >&3
  load $HOME/test/'test_helper/batsassert/load.bash'
 load $HOME/test/'test_helper/batscore/load.bash'
  load './one.sh'
}

```
it actually fails.... any idea? 

```
abc.sh 
process_one() {
echo "Hello "
}
```

**Comment by martin-schulze-vireso** [comment]
The tests run in a subshell of setup_file for test isolation reasons. This means, that only exported variables and functions are visible from setup_file to test. `setup` runs in the same shell and therefore does not suffer from this limitation.

**Comment by websolutions-hamburg** [comment]
Could it be that the documentation is incorrect in this case? The tutorial section says the following:
https://bats-core.readthedocs.io/en/stable/tutorial.html#avoiding-costly-repeated-setups

```bash
setup_file() {
    load 'test_helper/common-setup'
    _common_setup
    PORT=$(project.sh start-echo-server 2>&1 >/dev/null)
    export PORT
}

@test "server is reachable" {
    nc -z localhost "$PORT"
}
```

And `_common_setup` is defined as:

```bash
_common_setup() {
    load 'test_helper/bats-support/load'
    load 'test_helper/bats-assert/load'
    # get the containing directory of this file
    # use $BATS_TEST_FILENAME instead of ${BASH_SOURCE[0]} or $0,
    # as those will point to the bats executable's location or the preprocessed file respectively
    PROJECT_ROOT="$( cd "$( dirname "$BATS_TEST_FILENAME" )/.." >/dev/null 2>&1 && pwd )"
    # make executables in src/ visible to PATH
    PATH="$PROJECT_ROOT/src:$PATH"
}
```

I would therefore assume that it is ok and intended to load the libraries in the `setup_file` call.

**Comment by martin-schulze-vireso** [comment]
The tutorial is misleading in this case. The main task of `__common_setup` here is to provide `PATH` such that `project.sh` is  found. The helper libraries are never used in this case. I think the issues stems from the fact that `__common_setup` was originally intended to be used within `setup`.
